src/Module/FollowConfirm.php

```diff
@@ -0,0 +1,93 @@
+<?php
+namespace Friendica\Module;
+
+use Friendica\App;
+use Friendica\BaseModule;
+use Friendica\Core\L10n;
+use Friendica\Core\Logger;
+use Friendica\Core\Protocol;
+use Friendica\Database\DBA;
+use Friendica\Model\Contact;
+use Friendica\Protocol\Diaspora;
+use Friendica\Protocol\ActivityPub;
+use Friendica\Util\DateTimeFormat;
+
+class FollowConfirm extends BaseModule
+{
+	public static function post()
+	{
+		$a = self::getApp();
+
+		$uid = local_user();
+		if (!$uid) {
+			notice(L10n::t('Permission denied.') . EOL);
+			return;
+		}
+
+		$intro_id = intval(defaults($_POST, 'intro_id'  , 0));
+		$duplex   = intval(defaults($_POST, 'duplex'    , 0));
+		$cid      = intval(defaults($_POST, 'contact_id', 0));
+		$hidden   = intval(defaults($_POST, 'hidden'    , 0));
+
+		if (empty($cid)) {
+			notice(L10n::t('No given contact.') . EOL);
+			return;
+		}
+
+		Logger::log('Confirming follower with contact_id: ' . $cid);
+
+		$contact = DBA::selectFirst('contact', [], ['id' => $cid, 'uid' => $uid]);
```

Comment:
```suggestion
		$contact = Contact::getContactForUser($cid, $uid);
```

Reply:
There is no function ```getContactForUser```.

Reply:
https://github.com/friendica/friendica/blob/14d8c12130672bae9ff2e86dd9d7cf669b743939/src/Model/Contact.php#L251

Reply:
Ah okay. So it is new. I had looked at squeet.me where there is a more stable "develop" branch. I still don't see why we should create a new function for every different condition.

Reply:
The point is to abstract the database away in models. The `Contact` module doesn't need to know that the `contact` table even exists. It doesn't even need to know there's a database at all. All it needs to know is that Contact get be retrieved based on a contact ID and a user ID.

This way we can change the way we store information about contacts without having to change every single module. This may prove useful for adding support for PostgreSQL.

Reply:
For PostgreSQL this doesn't matter in any way. Most DBA functions will work without any changes. Only complicated queries (that can only be done via ```DBA::p``` and ```DBA::fetchFirst``` could possibly need some abstraction.

Reply:
In the perspective of dependence management, it would be marginally useful to remove the module dependence to the database entirely.

Instead of having:

> Module depends on:
> 1. Database
> 2. Model which depends on:
>    1. Database

We would just have:

> Module depends on:
> 1. Model which depends on:
>    1. Database.

Again, this simplification would mainly help with not caring if we use a database or anything else in the module, which may or may not be useful in our project.

Reply:
yep.
f.e. what about a "persistation layer", which isn't a DB. Like we store it just in the cache because we don't need to persist it all the time. That's the reason for persistance abstraction.

We could wait for such Use Case and than we refactor all classes, or we move all DBA-calls into Model classes and so we just have to mention all Model classes.

That's the reason behind it.

Reply:
I'm really for refactoring only parts that need to be refactored instead of having the need to create around 70 model classes for this "just in case" thing. I don't like having the need to create a huge overhead for some vaguely possible use case.

Reply:
hm .. It's more a architectural point of view .. not a "concrete need" .. As @MrPetovan said, it would reduce the dependency from 2 to 1 class, which **could** help in the future :-)

f.e. In the future, I'd like to introduce other DBs rather than PostgreSQL too, like InMemory DBs, or Cloud DBs, or Oracle/MSSQL in case Friendica grows a little bit. Like we do currently have 5 types of caching backends. And as far as I know they are all in use.

But I don't want to bug you ..